src/Generating/FinishGen.h

```diff
@@ -260,14 +260,24 @@ class cFinishGenSprinkleFoliage :
 	public cFinishGen
 {
 public:
-	cFinishGenSprinkleFoliage(int a_Seed) : m_Noise(a_Seed), m_Seed(a_Seed) {}
-
+	cFinishGenSprinkleFoliage(int a_Seed, int a_MaxCactusHeight, int a_MaxSugarcaneHeight):
+		m_Noise(a_Seed),
+		m_Seed(a_Seed),
+		m_MaxCactusHeight(a_MaxCactusHeight),
+		m_MaxSugarcaneHeight(a_MaxSugarcaneHeight)
+	{
+	}
 protected:
 	cNoise m_Noise;
 	int    m_Seed;
+	int    m_MaxCactusHeight;
+	int    m_MaxSugarcaneHeight;
+
+	/** Tries to place sugarcane at the coords specified, returns true if successful, updates the top variable (hence the & a_RefY) */
+	bool TryAddSugarcane(cChunkDesc & a_ChunkDesc, int a_RelX, HEIGHTTYPE & a_RelY, int a_RelZ);
 
-	/** Tries to place sugarcane at the coords specified, returns true if successful */
-	bool TryAddSugarcane(cChunkDesc & a_ChunkDesc, int a_RelX, int a_RelY, int a_RelZ);
+	/** Tries to place cactus at the coords specified, returns true if successful, updates the top variable (hence the & a_RefY) */
+	bool TryAddCactus(cChunkDesc & a_ChunkDesc, int a_RelX, HEIGHTTYPE & a_RelY, int a_RelZ);
```

Comment:
Could you just return the updated height rather than via an in-out parameter.  I think it's more readable that way.

Reply:
> Could you just return the updated height rather than via an in-out parameter. I think it's more readable that way.

I wouldn't agree, because you could forget to update the height (the function name doesn't imply to return a height. So calls like `TryAddCactus(chunk, relX, relY, relZ)` don't look wrong). 

But maybe some sort of transparent BlockSetter interface, which abstracts the height change away would maybe be the superior way:

```cpp
void TryAddSugarcane(Vector pos, BlockSetter & setter)
{
    setter.setBlockAt(1, 2, 3, E_STONE);
    setter.setBlockAt(pos, E_SUGARCANE);
}
```

Reply:
>I wouldn't agree, because you could forget to update the height

You could equally well forget to update the height returned by the out param.  You still need to call `a_ChunkDesc.SetHeight` either way.

In actual fact, why not just update the heightmap in the `TryPlace` function.

Reply:
> In actual fact, why not just update the heightmap in the TryPlace function.

I don't see, where `TryPlace` is being used. I would have used the `SetBlockType` function.

Reply:
I meant `TryPlaceCactus` and `TryPlaceSugarcane`.

Reply:
Yeah, that would also be better. Is there any reason, not to put the heightmap thing into `SetBlockType`, etc. functions?

Reply:
I would assume it's to avoid doing unnecessary work.  It would be interesting to see a profile/benchmark though.

Reply:
Indeed, I think putting it in the `TryPlace` functions is a better solution - I had kept the boolean return value in case making checks to see if a sugarcane or a cacti are placeable was ever needed.